lib/spack/spack/multimethod.py

```diff
@@ -111,14 +108,26 @@ def __call__(self, package_self, *args, **kwargs):
             return self.default(package_self, *args, **kwargs)
 
         else:
-            superclass = super(package_self.__class__, package_self)
```

Comment:
The method described in the stackoverflow link using python 3 would not work in this case, because it relies on the class being implicitly passed by the scope, which it isn't in this case. If we used that, we would get the superclass of `SpecMultiMethod`, not the superclass of `package_self`. Similarly, we can't avoid using `package_self.__class__`, because we cannot statically know what that value will be, as it is different for every package. Otherwise you are absolutely correct that this is the same bug discussed on stackoverflow and mrsenko. 

I agree that I should change the comment to reflect the fact that `super()` returns, rather than uses, a proxy object.